File: src/main/java/org/modelingvalue/nelumbo/patterns/TokenTypePattern.java
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// (C) Copyright 2018-2026 Modeling Value Group B.V. (http://modelingvalue.org)                                        ~
//                                                                                                                     ~
// Licensed under the GNU Lesser General Public License v3.0 (the 'License'). You may not use this file except in      ~
// compliance with the License. You may obtain a copy of the License at: https://choosealicense.com/licenses/lgpl-3.0  ~
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on ~
// an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the  ~
// specific language governing permissions and limitations under the License.                                          ~
//                                                                                                                     ~
// Maintainers:                                                                                                        ~
//     Wim Bast, Tom Brus                                                                                              ~
//                                                                                                                     ~
// Contributors:                                                                                                       ~
//     Victor Lap                                                                                                      ~
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

package org.modelingvalue.nelumbo.patterns;

import java.io.Serial;

import org.modelingvalue.collections.List;
import org.modelingvalue.collections.mutable.MutableList;
import org.modelingvalue.nelumbo.AstElement;
import org.modelingvalue.nelumbo.Node;
import org.modelingvalue.nelumbo.Type;
import org.modelingvalue.nelumbo.Variable;
import org.modelingvalue.nelumbo.syntax.ParseState;
import org.modelingvalue.nelumbo.syntax.Token;
import org.modelingvalue.nelumbo.syntax.TokenType;

public class TokenTypePattern extends Pattern {
    @Serial
    private static final long serialVersionUID = 2405616043878166113L;

    public TokenTypePattern(Type type, List<AstElement> elements, Object... args) {
        super(type, elements, args);
    }

    protected TokenTypePattern(Object[] args, TokenTypePattern declaration) {
        super(args, declaration);
    }

    @Override
    protected TokenTypePattern struct(Object[] array, Node declaration) {
        return new TokenTypePattern(array, (TokenTypePattern) declaration);
    }

    public TokenType tokenType() {
        return (TokenType) get(0);
    }

    @Override
    public String toString(TokenType[] previous) {
        return "<" + tokenType() + ">";
    }

    @Override
    public ParseState state(ParseState next, Functor functor) {
        return new ParseState(tokenType(), next);
    }

    @Override
    public List<Type> argTypes(List<Type> types) {
        TokenType type = tokenType();
        return !isEmpty(type) ? types.add(Type.$STRING) : types;
    }

    @Override
    protected int string(List<Object> args, int ai, StringBuffer sb, TokenType[] previous, boolean alt) {
        TokenType type = tokenType();
        if (!isEmpty(type)) {
            Object val = args.get(ai);
            if (val instanceof String text && type.matches(text)) {
                addText(sb, previous, text);
                return ai + 1;
            } else {
                String text = val != null ? val.toString() : null;
                if (text != null && type.matches(text)) {
                    addText(sb, previous, text);
                    return ai + 1;
                } else {
                    return -1;
                }
            }
        }
        return ai;
    }

    @Override
    protected int args(List<AstElement> elements, int i, MutableList<Object> args, boolean alt, Functor functor) {
        if (i < elements.size()) {
            AstElement e = elements.get(i);
            TokenType type = tokenType();
            if (e instanceof Token t) {
                if (t.isKeyword() && type.isVariableContent()) {
                    return -1;
                } else if (t.type().equals(type)) {
                    if (!isEmpty(type)) {
                        args.add(t.text());
                    }
                    return i + 1;
                } else if (TokenType.NEWLINE.equals(type) && Pattern.isEndOfLine(t)) {
                    return i;
                }
            } else if (e instanceof Variable v && type.equals(v.type().tokenType())) {
                args.add(v);
                return i + 1;
            }
        }
        return -1;
    }

    private static boolean isEmpty(TokenType type) {
        return type == TokenType.NEWLINE || type == TokenType.BEGINOFFILE || type == TokenType.ENDOFFILE || type == TokenType.ENDOFLINE;
    }

    @Override
    public Pattern declaration(Token token) {
        return null;
    }

}

File: web-ts/src/patterns/TokenTypePattern.ts
/**
 * TokenTypePattern - matches a specific token type.
 * Ported from Java: org.modelingvalue.nelumbo.patterns.TokenTypePattern
 */

import { List, Map } from 'immutable';
import { TokenType } from '../syntax/TokenType';
import { Token } from '../syntax/Token';
import type { AstElement } from '../AstElement';
import { Type } from '../Type';
import { Variable } from '../Variable';
import { Node } from '../Node';
import { Pattern } from './Pattern';
import { ParseState } from '../syntax/ParseState';
import type { Functor } from './Functor';

export class TokenTypePattern extends Pattern {
  constructor(type: Type, elements: List<AstElement>, ...args: unknown[]) {
    super(type, elements, ...args);
  }

  protected static fromData(data: unknown[], declaration?: Node): TokenTypePattern {
    const pattern = Object.create(TokenTypePattern.prototype) as TokenTypePattern;
    (pattern as unknown as { _data: unknown[] })._data = data;
    (pattern as unknown as { _declaration: Node })._declaration = declaration ?? pattern;
    return pattern;
  }

  protected struct(data: unknown[], declaration?: Node): TokenTypePattern {
    return TokenTypePattern.fromData(data, declaration ?? this.declaration());
  }

  tokenType(): TokenType {
    return this.get(0) as TokenType;
  }

  toString(_previous?: TokenType[]): string {
    return '<' + this.tokenType().name + '>';
  }

  parseState(next: ParseState, _functor: Functor): ParseState {
    return new ParseState(this.tokenType(), next);
  }

  argTypes(types: List<Type>): List<Type> {
    const tt = this.tokenType();
    return !this.isEmpty(tt) ? types.push(Type.$STRING) : types;
  }

  string(args: List<unknown>, ai: number, sb: string[], previous: TokenType[], _alt: boolean): number {
    const tt = this.tokenType();
    if (!this.isEmpty(tt)) {
      const val = args.get(ai);
      let text: string | null = null;

      if (typeof val === 'string') {
        text = val;
      } else if (val !== null && val !== undefined) {
        text = String(val);
      }

      if (text !== null && this.matches(tt, text)) {
        this.addText(sb, previous, text);
        return ai + 1;
      }
      return -1;
    }
    return ai;
  }

  private matches(tt: TokenType, text: string): boolean {
    // Check if the text matches this token type pattern
    if (tt.pattern) {
      const match = text.match(tt.pattern);
      return match !== null && match[0] === text;
    }
    return false;
  }

  extractArgs(
    elements: List<AstElement>,
    i: number,
    args: unknown[],
    _alt: boolean,
    _functor: Functor,
    _typeArgs: Map<Variable, Type>
  ): number {
    if (i < elements.size) {
      const e = elements.get(i);
      const tt = this.tokenType();

      if (e instanceof Token) {
        if (e.isTextMatch && tt.isVariableContent()) {
          return -1;
        } else if (e.type === tt) {
          if (!this.isEmpty(tt)) {
            args.push(e.text);
          }
          return i + 1;
        } else if (TokenType.NEWLINE === tt && Pattern.isEndOfLine(e)) {
          return i;
        }
      } else if (e instanceof Variable && tt === e.type().tokenType()) {
        args.push(e);
        return i + 1;
      }
    }
    return -1;
  }

  private isEmpty(tt: TokenType): boolean {
    return tt === TokenType.NEWLINE ||
           tt === TokenType.BEGINOFFILE ||
           tt === TokenType.ENDOFFILE ||
           tt === TokenType.ENDOFLINE;
  }

  tokenDeclaration(_token: Token): Pattern | null {
    return null;
  }
}
